index.html

```diff
@@ -215,40 +215,6 @@ <h2>
             </ol>
           </li>
```

Comment:
I am not sure we want to show any permission prompt without user activation @reillyeon ?

Reply:
But actually raphael, I am not sure it always have to be gates behind a user activation. It the user accepted it on a previous visit, it is auto granted and thus no user activation is needed. But if we want to show the prompt ("prompt" state) well then we probably need a user activation to do so

Reply:
In the WebNFC API we use custom logic to use the "query" permission algorithm when there isn't user activation and the "request" algorithm when there is. In contrast in the Idle Detection API we always use the "query" algorithm and provide a separate method which only requests permission and thus requires user activation.

Either approach is fine by me though the whole discussion is somewhat moot because Chromium doesn't implement the "screen-wake-lock" permission. It is always granted.

Reply:
> In the WebNFC API we use custom logic to use the "query" permission algorithm when there isn't user activation and the "request" algorithm when there is.

From a quick look, the [spec](https://w3c.github.io/web-nfc/#dfn-obtain-permission) has a custom "obtain permission" algorithm that mimics "request permission" fairly closely and does not mention user activation anywhere, while the Chromium implementation always just calls its "request permission" equivalent and always blocks requests that were not user-activated. w3c/web-nfc#482 then references the explainer, which says users should query the current permission state and trigger "obtain permission" via a user-activated action. This mismatch sounds confusing to me.

> In contrast in the Idle Detection API we always use the "query" algorithm and provide a separate method which only requests permission and thus requires user activation.

We used to do that, but decided to remove `requestPermission()` in #237.

Having said that, I guess the question is whether it is worth implementing a custom "request permission" algorithm that checks for user-activation somewhere in the steps (be it to always check for it upfront or only when the permission state is "prompt"), or to hope w3c/permissions#189 and/or w3c/permissions#194 are solved at some point so we can rely on the Permissions spec to take care of it.

Reply:
In the absence of support from the Permissions API specification I think we should define custom algorithms but try to make sure they are consistent between specifications that the desire the "request permission if there is user activation" behavior. This can form the draft of the language that will hopefully be upstreamed at some point.

At its core the WebNFC issue seems to highlight the concern that requiring user activation should not be specified normatively and browsers can make that determination on their own, citing differences between Chrome and Firefox. In that case it would seem that always using the default "request" algorithm is fine and the user activation requirement is an optional intervention that implementations may choose to implement for some or all permissions.

I personally feel that requiring user activation for permission prompts should be normatively specified so I support adding the necessary custom steps here.

Reply:
> I personally feel that requiring user activation for permission prompts should be normatively specified so I support adding the necessary custom steps here.

Can you elaborate on what you have in mind? We can either just fine-tune the existing algorithm (by e.g. simplifying the query part) or do something like checking for user activation first and then doing the rest.

Reply:
One other thing comes to mind: if we do not take the current Chromium implementation of this spec into consideration for a moment, doesn't conditionally checking for user activation only when the permission state is "prompt" lead to API users to coding with this scenario in mind, basically making `WakeLock.request()` act as API unconditionally gated by user-activation?

Reply:
The proposed logic for an API which requires a permission and will request it if it hasn't been granted is:

Query for permission "foo", then
1. If "allow", continue.
2. If "deny", fail.
3. If "prompt", then
    1. If no user activation, fail.
    2. Request permission "foo", then
        1. If "allow", continue.
        2. If "deny", fail.

This doesn't mean that the method unconditionally requires user activation. It only means it require user activation if permission has not already been granted.

Reply:
Yeah, that kind of matches what we already have and which we can fine-tune, my concern with this approach is the one I mentioned above: wouldn't API authors either need to write code that looks like w3c/web-nfc#482 or always call `WakeLock.request()` via user activation since they don't know whether it will work without it, in practice acting as if the algorithm was always gated by user activation?

Reply:
If asking for permission requires user activation then, whether there is a separate `requestPermission()` method or not, for developers the process of requesting a wake lock requires a user gesture if they can't assume they have permission. That seems to be working as intended. Applications which know they already have permission (perhaps because they asked for permission in the past) don't need a user gesture.

For example, a site might add a button to enable the wake lock (and thus have activation while making the initial request) but call `request()` without a user gesture in the `visibilitychanged` handler to reacquire the lock later.

I'd like to see a resolution to the issues raised on the WebNFC and Permissions APIs but I don't think that should block this change.

Reply:
(for the record, while we discuss this I've submitted #303 anyway to clean up the algorithm we currently have)

Reply:
Will you file issues in Web NFC and permission repos ?

Reply:
Who's "you" in this case and what issues are you referring to? :-) I've mentioned 2 issues in the Permissions spec and 1 in the Web-NFC spec, so it's not clear what other issues still need to be filed.

Reply:
yes, I see now. Sorry have only half been following this discussion from the side lines :)